From 7b78e5daae761c8fb9b4637da42f61f2ccd58a4f Mon Sep 17 00:00:00 2001
From: Vasily Khoruzhick <anarsoul@gmail.com>
Date: Fri, 10 Jan 2020 19:33:21 -0800
Subject: [PATCH 046/102] lima: fix PLBU_CMD_PRIMITIVE_SETUP command

Apparently it doesn't depend on primitive type, the value
only depends on whether we specify point size via PLBU command --
bit 12 is set in this case

Signed-off-by: Vasily Khoruzhick <anarsoul@gmail.com>
---
 src/gallium/drivers/lima/lima_draw.c | 17 ++++++++++-------
 1 file changed, 10 insertions(+), 7 deletions(-)

diff --git a/src/gallium/drivers/lima/lima_draw.c b/src/gallium/drivers/lima/lima_draw.c
index e59f46bdee6..187a4f868c5 100644
--- a/src/gallium/drivers/lima/lima_draw.c
+++ b/src/gallium/drivers/lima/lima_draw.c
@@ -148,8 +148,9 @@ struct lima_render_state {
 #define PLBU_CMD_VIEWPORT_TOP(v) PLBU_CMD(v, 0x10000106)
 #define PLBU_CMD_ARRAYS_SEMAPHORE_BEGIN() PLBU_CMD(0x00010002, 0x60000000)
 #define PLBU_CMD_ARRAYS_SEMAPHORE_END() PLBU_CMD(0x00010001, 0x60000000)
-#define PLBU_CMD_PRIMITIVE_SETUP(prim, cull, index_size) \
-   PLBU_CMD(0x200 | (prim) | (cull) | ((index_size) << 9), 0x1000010B)
+#define PLBU_CMD_PRIMITIVE_SETUP(force_point_size, cull, index_size) \
+   PLBU_CMD(0x2200 | (force_point_size ? 0x1000 : 0) | \
+            (cull) | ((index_size) << 9), 0x1000010B)
 #define PLBU_CMD_RSW_VERTEX_ARRAY(rsw, gl_pos) \
    PLBU_CMD(rsw, 0x80000000 | ((gl_pos) >> 4))
 #define PLBU_CMD_SCISSORS(minx, maxx, miny, maxy) \
@@ -853,12 +854,14 @@ lima_pack_plbu_cmd(struct lima_context *ctx, const struct pipe_draw_info *info)
          cull |= ccw ? 0x00020000 : 0x00040000;
    }
 
-   if (info->mode == PIPE_PRIM_POINTS && ctx->vs->point_size_idx != -1)
-      PLBU_CMD_PRIMITIVE_SETUP(0x0000, cull, info->index_size);
-   else if (info->mode < PIPE_PRIM_TRIANGLES)
-      PLBU_CMD_PRIMITIVE_SETUP(0x3000, cull, info->index_size);
+   if ((ctx->vs->point_size_idx != -1 && info->mode == PIPE_PRIM_POINTS) ||
+       info->mode >= PIPE_PRIM_TRIANGLES)
+      PLBU_CMD_PRIMITIVE_SETUP(false, cull, info->index_size);
    else
-      PLBU_CMD_PRIMITIVE_SETUP(0x2000, cull, info->index_size);
+      /* Point size or line width needs to be specified with PLBU command if
+       * shader doesn't write it
+       */
+      PLBU_CMD_PRIMITIVE_SETUP(true, cull, info->index_size);
 
    PLBU_CMD_RSW_VERTEX_ARRAY(
       lima_ctx_buff_va(ctx, lima_ctx_buff_pp_plb_rsw, LIMA_CTX_BUFF_SUBMIT_PP),
-- 
2.24.1

